### PR TITLE
DATAREDIS-315 - Add support for Redis Cluster.

### PR DESCRIPTION
**Pull request by christophstrobl**
**NOTE:** Redis Cluster is currently only supported for https://github.com/xetorthio/jedis and https://github.com/mp911de/lettuce.

Cluster support is based on the very same building blocks as non clustered communication. `RedisClusterConnection` and extension to `RedisConnection` handles the communication with the Redis Cluster and translates errors into the Spring DAO exception hierarchy.
`RedisClusterConnection` 's are created via the `RedisConnectionFactory` which has to be set up with the according `RedisClusterConfiguration`.

```java
@Configuration
public class AppConfig {

    /*
     * spring.redis.cluster.nodes[0] = 127.0.0.1:7379
     * spring.redis.cluster.nodes[1] = 127.0.0.1:7380
     * ...
     */
    @Value("${spring.redis.cluster.nodes}")
    private Collection<String> initialClusterNodes;

    public @Bean RedisConnectionFactory connectionFactory() {

        return new JedisConnectionFactory(
            new RedisClusterConfiguration(initialClusterNodes));
    }
}
```

Redis Cluster behaves slightly different from single node Redis or even a Sentinel monitored master slave environment. This is reasoned by the automatic sharding that maps a key to one of 16384 slots which are distributed across the nodes. Therefore commands that involve more than one key must assert that all keys map to the exact same slot in order to avoid crosslot execution errors.
Further on, hence a single cluster node, only servers a dedicated set of keys, commands issued against one particular server only return results for those keys served by the server. As a very simple example take the `KEYS` command. When issued to a server in cluster environment it only returns the keys served by the node the request is sent to and not necessarily all keys within the cluster. So to get all keys in cluster environment it is necessary to read the keys from at least all known master nodes.

While redirects for to a specific keys to the corresponding slot serving node are handled by the driver libraries, higher level functions like collecting information across nodes, or sending commands to all nodes in the cluster that are covered by `RedisClusterConnection`. Picking up the keys example from just before, this means, that the `keys(pattern)` method picks up every master node in cluster and simultaneously executes the `KEYS` command on every single one, while picking up the results and returning the cumulated set of keys. To just request the keys of a single node `RedisClusterConnection` provides overloads for those (like `keys(node, pattern)` ). 

Crosslot requests such as `MEGT` are automatically served by the native driver library when all keys map to the same slot. However once this is not the case `RedisClusterConnection` executes multiple parallel `GET` commands against the slot serving nodes and again returns a cumulated result. Obviously this is less performing than the single slot execution and therefore should be used with care. In doubt please consider pinning keys to the same slot by providing a prefix in curly brackets like `{my-prefix}.foo` and `{my-prefix}.bar` which will both map to the same slot number.

`RedisTemplate` provides access to cluster specific operations via the `ClusterOperations` interface that can be obtained via `RedisTemplate.opsForCluster()`. This allows to execute commands explicitly on a single node within the cluster while retaining de-/serialization features configured for the template and provides administrative commands such as `CLUSTER MEET` or more high level operations for eg. resharding.

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -1,14 +1,14 @@
 slf4jVersion=1.7.12
 junitVersion=4.12
 jredisVersion=06052013
-jedisVersion=2.7.3
+jedisVersion=2.8.0
 springVersion=4.1.7.RELEASE
 springDataBuildVersion=1.8.0.BUILD-SNAPSHOT
 log4jVersion=1.2.17
-version=1.7.0.BUILD-SNAPSHOT
+version=1.7.0.DATAREDIS-315-SNAPSHOT
```
Comment (mp911de) [inline gradle.properties]: You sure about `DATAREDIS-315-SNAPSHOT`?

File: src/asciidoc/reference/redis-cluster.adoc [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+<5> Request routed to node at 7381 serving slot 12182
+<6> Request routed to node at 7379 serving slot 5061
+<7> Request routed to nodes at 7379, 7380, 7381 -> [foo, bar]
+<8> Request routed to node at 7379 -> [bar]
+<9> Request routed to node at 7380 -> []
+<10> Request routed to node at 7381 -> [foo]
+<11> Request routed to node at 7382 -> [bar]
+====
+
+Crosslot requests such as `MEGT` are automatically served by the native driver library when all keys map to the same slot. However once this is not the case `RedisClusterConnection` executes multiple parallel `GET` commands against the slot serving nodes and again returns a cumulated result. Obviously this is less performing than the single slot execution and therefore should be used with care. In doubt please consider pinning keys to the same slot by providing a prefix in curly brackets like `{my-prefix}.foo` and `{my-prefix}.bar` which will both map to the same slot number.
```
Comment (mp911de) [inline src/asciidoc/reference/redis-cluster.adoc]: s/MEGT/MGET

File: src/asciidoc/reference/redis-cluster.adoc [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+.Sample of Crosslot Request Handling
```
Comment (mp911de) [inline src/asciidoc/reference/redis-cluster.adoc]: s/Crosslot/Cross Slot

File: gradle.properties (hunk 1 above)
Comment (christophstrobl) [inline gradle.properties]: @mp911de I am :smile: . We normally bump the version in the first commit of an issue branch to have dedicated build artifacts. That commit is dropped when merging into master.

**Comment by zhengfc** [comment]
Hi, I used 1.7.0.DATAREDIS-315-SNAPSHOT project. I found a problem, when a master is shutdown, but this time a slave change master, the redis cluster stat is ok. When it happen the previous master is also master node, just flag is fail, will cause some wrong. I suggest the RedisClusterNode add the flag or ClusterTopology getMasterNodes() just return masters(slots is not null).

**Comment by christophstrobl** [comment]
@zhengfc thanks for the heads up - I'll add checks on `flag` and `link-state` to handle nodes like the one below:

```text
b8b5ee... 127.0.0.1:7382 master,fail - 1450160048933 1450160048832 38 disconnected
```

**Comment by shameertk** [comment]
When we tried connecting to redis cluster via redis.clients.jedis.JedisCluster it's working. Same cluster when we tried using spring data DATAREDIS-315 branch we are getting below error. Could you add a Junit test case to set & get a key value pair using RedisClusterConfiguration

System.out.println(" redisTemplate.conn.ping() : " + redisTemplate.getConnectionFactory().getConnection().ping(
) );

Exception :  
org.springframework.data.redis.ClusterStateFailureExeption: Could not retrieve cluster info.

**Comment by christophstrobl** [comment]
@shameertk SD-Redis uses `CLUSTER  NODES` command for reading cluster topology by running over the set of available connections `JedisCluster.getClusterNodes()` provides. On first valid result that one is returned. If none returns a result the error is raised.
Could you provide a bit more insight into your actual cluster topology and java setup? I definitely need to improve the error message there.

**Comment by farhan-rayani** [comment]
Hi Christoph,

My colleague Shameer had pointed out the issue we are facing. On reply of this I created the test project from this example. Kindly find it here:

https://github.com/flydubai-IT/spring-redis-cluster

On this I had implemented the RedisTemplate on top of JedisConnectionfactory. Tried the single node redis connection working fine (SingleJRedisTest.java). Tried to set the data in cluster node not working (JRedisTest.java). Kindly run this class JRedisTest.java and let me know if I am missing something. Had hard coded the cluster nodes IPs in the config classes (AppSingleConfig.java & AppConfig.java) feel free to change them.Thanks.

**Comment by christophstrobl** [comment]
Thanks @farhan-rayani, @shameertk 
I've been checking the sample against my local installation. Despite IPs I changed the following, which made it work.

* Upgrade to **jedis 2.8** (required since jedis 2.7 lacks support for binary commands on `JedisCluster`).
```xml
<dependency>
  <groupId>redis.clients</groupId>
  <artifactId>jedis</artifactId>
  <version>2.8.0</version>
</dependency>
```

* Use JUnit to run test case.
```java
@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration(classes = { AppConfig.class }) // app config with cluster env parameters
public class SingleJRedisTest {

  @Autowired RedisTemplate<String, String> redisTemplate;

  @Before
  public void setUp() {

    // use a clean instance | remove if you want to keep values.
    redisTemplate.execute(new RedisCallback<String>() {

      public String doInRedis(RedisConnection connection) throws DataAccessException {
        connection.flushDb();
        return "DONE";
      }
    });
  }

  @Test
  public void main() {  // the previous static void main(String[] ar... method

    // CLUSTER KEYSLOT "name" => 5798
    redisTemplate.opsForValue().set("name", "farhan");

    System.out.println(redisTemplate.opsForValue().get("name"));
    assertThat(redisTemplate.opsForValue().get("name"), is("farhan"));
  }
}
```

**Comment by farhan-rayani** [comment]
Hi Christoph,

Thanks for your quick reply.

I tried still getting exception had checked the redis clusters manually its fine. Kindly run Clustertest.java. There is one more class TestCluster.java which is not using Spring Data and its working fine. Let me know if I am missing something.

Kindly take the update from https://github.com/flydubai-IT/spring-redis-cluster

org.springframework.data.redis.ClusterStateFailureExeption: Could not retrieve cluster info.
	at org.springframework.data.redis.connection.jedis.JedisClusterConnection$JedisClusterTopologyProvider.getTopology(JedisClusterConnection.java:3810)
	at org.springframework.data.redis.connection.ClusterCommandExecutor.getClusterTopology(ClusterCommandExecutor.java:291)
	at org.springframework.data.redis.connection.ClusterCommandExecutor.executeCommandOnAllNodes(ClusterCommandExecutor.java:152)
	at org.springframework.data.redis.connection.jedis.JedisClusterConnection.flushDb(JedisClusterConnection.java:2727)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.springframework.data.redis.core.CloseSuppressingInvocationHandler.invoke(CloseSuppressingInvocationHandler.java:57)
	at com.sun.proxy.$Proxy13.flushDb(Unknown Source)
	at com.flydubai.security.server.test.ClusterTest$1.doInRedis(ClusterTest.java:29)
	at com.flydubai.security.server.test.ClusterTest$1.doInRedis(ClusterTest.java:1)
	at org.springframework.data.redis.core.RedisTemplate.execute(RedisTemplate.java:191)
	at org.springframework.data.redis.core.RedisTemplate.execute(RedisTemplate.java:153)
	at org.springframework.data.redis.core.RedisTemplate.execute(RedisTemplate.java:141)
	at com.flydubai.security.server.test.ClusterTest.setUp(ClusterTest.java:26)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:83)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:72)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:231)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:88)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:71)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:174)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:50)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)

**Comment by christophstrobl** [comment]
@farhan-rayani sd-redis 1.7 requires spring 4.1 - I noticed some 3.2 version in you pom.xml. futher I just pushed some code to get a more detailed error message for loading cluster information - it might take a while till the artifact is available via repo.spring.io

**Comment by zhengfc** [comment]
Hi christophstrobl,
Hope SDR-315 join READONY from slave. This will enables read queries for a connection to a Redis Cluster slave node. 
Thank you very much
http://redis.io/commands/readonly

**Comment by christophstrobl** [comment]
@zhengfc `READONLY` is actually a bit tricky and might be handled in a subsequent PR.

**Comment by zhengfc** [comment]
@christophstrobl Thanks for your quick reply.
I will keep watch the READONLY question. Expect message about read and write seperate :smile:

File: src/main/java/org/springframework/data/redis/connection/ClusterCommandExecutor.java [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+
+			done = true;
+			for (Map.Entry<RedisClusterNode, Future<T>> entry : futures.entrySet()) {
+
+				if (!entry.getValue().isDone() && !entry.getValue().isCancelled()) {
+					done = false;
+				} else {
+					if (!result.containsKey(entry.getKey()) && !exceptions.containsKey(entry.getKey())) {
+						try {
+							result.put(entry.getKey(), entry.getValue().get());
```
Comment (mp911de) [inline src/main/java/org/springframework/data/redis/connection/ClusterCommandExecutor.java]: Awaiting a `Future` without a timeout can block forever. You get here only if the `Future` is done but the outer loop is also not protected by a timeout.

File: src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConverters.java [hunk 5]
```diff
@@ -199,6 +208,74 @@ public Tuple convert(ScoredValue<byte[]> source) {
...
+		};
+
+		CLUSTER_NODE_TO_CLUSTER_NODE_CONVERTER = new Converter<com.lambdaworks.redis.cluster.models.partitions.RedisClusterNode, RedisClusterNode>() {
+
+			@Override
+			public RedisClusterNode convert(com.lambdaworks.redis.cluster.models.partitions.RedisClusterNode source) {
+
+				Set<Flag> flags = parseFlags(source.getFlags());
+
+				return RedisClusterNode.newRedisClusterNode().listeningAt(source.getUri().getHost(), source.getUri().getPort())
```
Comment (mp911de) [inline src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConverters.java]: Lettuce returns `null` for RedisURI if the cluster node has no address (`noaddr` flag, see https://github.com/mp911de/lettuce/issues/173)

File: src/main/java/org/springframework/data/redis/connection/RedisNode.java [hunk 6]
```diff
@@ -121,11 +182,21 @@ public RedisNodeBuilder() {
...
+		/**
+		 * Set host and port of server.
+		 * 
+		 * @param host must not be {@literal null}.
+		 * @param port
+		 * @return
+		 */
 		public RedisNodeBuilder listeningAt(String host, int port) {
 
 			Assert.hasText(host, "Hostname must not be empty or null.");
```
Comment (mp911de) [inline src/main/java/org/springframework/data/redis/connection/RedisNode.java]: `host` can be empty (`null` RedisURI in lettuce) in a cluster when the node has the `noaddr` flag

File: src/main/java/org/springframework/data/redis/connection/jedis/JedisConverters.java [hunk 7]
```diff
@@ -200,6 +225,21 @@ public static String toString(byte[] source) {
...
+	 * @return
+	 * @since 1.7
+	 */
+	public static Set<RedisClusterNode> toSetOfRedisClusterNodes(String clusterNodes) {
+
+		if (StringUtils.isEmpty(clusterNodes)) {
+			return Collections.emptySet();
+		}
+
+		String[] lines = clusterNodes.split(System.getProperty("line.separator"));
```
Comment (mp911de) [inline src/main/java/org/springframework/data/redis/connection/jedis/JedisConverters.java]: The line separator in `CLUSTER NODES` is not platform-dependent, it's always `\n` (https://github.com/antirez/redis/blob/unstable/src/cluster.c#L3747)

File: src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnectionFactory.java [hunk 8]
```diff
@@ -419,8 +501,18 @@ public boolean isRedisSentinelAware() {
...
+	 */
+	public boolean isClusterAware() {
+		return clusterConfiguration != null;
+	}
+
 	@Override
 	public RedisSentinelConnection getSentinelConnection() {
-		return new LettuceSentinelConnection(client.connectSentinelAsync());
+		if (!(client instanceof RedisClient)) {
+			throw new InvalidDataAccessResourceUsageException("Unable to connect to senitels using " + client.getClass());
```
Comment (mp911de) [inline src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnectionFactory.java]: s/senitels/sentinels

File: src/main/java/org/springframework/data/redis/core/ClusterOperations.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+	/**
+	 * Add slots in {@link SlotRange} to given node.
+	 * 
+	 * @param node must not be {@literal null}.
+	 * @param range must not be {@literal null}.
+	 */
+	void addSlots(RedisClusterNode node, SlotRange range);
+
+	/**
+	 * tart an {@literal Append Only File} rewrite process on given node.
```
Comment (mp911de) [inline src/main/java/org/springframework/data/redis/core/ClusterOperations.java]: s/tart/Start

File: src/test/java/org/springframework/data/redis/connection/ClusterCommandExecutorUnitTests.java [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+	static final RedisClusterNode CLUSTER_NODE_2 = RedisClusterNode.newRedisClusterNode()
+			.listeningAt(CLUSTER_NODE_2_HOST, CLUSTER_NODE_2_PORT).serving(new SlotRange(5461, 10922))
+			.withId("0f2ee5df45d18c50aca07228cc18b1da96fd5e84").promotedAs(NodeType.MASTER).linkState(LinkState.CONNECTED)
+			.build();
+	static final RedisClusterNode CLUSTER_NODE_3 = RedisClusterNode.newRedisClusterNode()
+			.listeningAt(CLUSTER_NODE_3_HOST, CLUSTER_NODE_3_PORT).serving(new SlotRange(10923, 16383))
+			.withId("3b9b8192a874fa8f1f09dbc0ee20afab5738eee7").promotedAs(NodeType.MASTER).linkState(LinkState.CONNECTED)
+			.build();
+
+	static final RedisClusterNode UNKNOWN_CLUSTER_NODE = new RedisClusterNode("8.8.8.8", 7379, null);
```
Comment (mp911de) [inline src/test/java/org/springframework/data/redis/connection/ClusterCommandExecutorUnitTests.java]: Poor Google DNS :-)

**Comment by mp911de** [comment]
Merged.

**Comment by oliu** [comment]
hi,I use spring-data-redis 1.7.0M1 and jedis use 2.8.0.when I test,there has exception like org.springframework.data.redis.ClusterRedirectException: Redirect: slot 12182 to 192.168.1.223:7002.; nested exception is redis.clients.jedis.exceptions.JedisMovedDataException: MOVED 12182 192.168.1.223:7002
at org.springframework.data.redis.connection.jedis.JedisExceptionConverter.convert(JedisExceptionConverter.java:56)
at org.springframework.data.redis.connection.jedis.JedisExceptionConverter.convert(JedisExceptionConverter.java:41)

**Comment by christophstrobl** [comment]
@oliu please provide information about your actual `RedisConnectionFactory` setup and the full stacktrace. You can have a look at the [reference documentation](http://docs.spring.io/spring-data/redis/docs/1.7.0.M1/reference/html/#cluster) as well as an [example project](https://github.com/spring-projects/spring-data-examples/tree/master/redis/cluster) for information on cluster config for sd-redis. Feel free to open an issue in in the [bug tracker](https://jira.spring.io/browse/DATAREDIS).

**Comment by oliu** [comment]
@christophstrobl Thank you vary much.It is ok now.It is my cluster config problem.

**Pull request closed without merging**